### DISCUSSION TITLE
Pika confirm_delivery not accepting callback argument

### DISCUSSION THREAD

**Discussion by margaretli27** [pika/pika#1415] [Q&A]
Based on the [method signature](https://github.com/pika/pika/blob/e23ddd5be89ae8712912d4b323db8619ed100de3/pika/channel.py#L555) of confirm_delivery, I expect to be able to set two different callback parameters. However when I try that with pika 1.3.1 (or 1.3.0), with `channel.confirm_delivery(on_delivery_confirmation)` where `on_delivery_confirmation` is my callback, I get
```
TypeError: confirm_delivery() takes 1 positional argument but 2 were given
```

I inspected the method signature directly in my terminal and saw the two parameters I expect were absent. 
```
>>> inspect.getfullargspec(channel.confirm_delivery)
FullArgSpec(args=['self'], varargs=None, varkw=None, defaults=None, kwonlyargs=[], kwonlydefaults=None, annotations={})
```

But in the source code they are clearly there. Am I missing something really obvious with using this method, or did the parameters get sneakily removed at some point? Appreciate any guidance here

**Comment by lukebakken** [comment]
Hello! Please share *all* of your code. Ideally, I should be able to clone a git repository and run it.

The reason I ask is that by having your code I don't have to guess.

**Reply by margaretli27** [reply]
Hi @lukebakken thanks for responding. There's not really any code to share; I just got started with pika so have essentially just the below
```
import pika

def on_delivery_confirmation(method_frame):
    print(method_frame)

parameters = pika.URLParameters(...)
connection = pika.BlockingConnection(parameters)
channel = connection.channel()
channel.confirm_delivery(on_delivery_confirmation)
```
I won't be able to get past this line. If this isn't enough I can upload a repo

**Reply by lukebakken** [reply]
OK, I see what's going on. Since you're using `BlockingConnection`, Pika takes care of the confirmations behind the scenes (https://github.com/lukebakken/pika-1415).

If you have low message publishing rates, this is fine. Otherwise, you should use `SelectConnection` and handle confirmations asynchronously:

https://github.com/pika/pika/blob/main/examples/confirmation.py

Let me know if you have any other questions!

**Reply by margaretli27** [reply]
I see, thanks! Does the delivery tag in the `frame` increment in perpetuity? I would like publish a batch of messages and infer from that tag which messages can be acked/nacked. I see [this example](https://github.com/pika/pika/blob/e23ddd5be89ae8712912d4b323db8619ed100de3/examples/asynchronous_publisher_example.py#L242) of how to ack/nack multiple messages, where [the publisher tracks an incrementing number](https://github.com/pika/pika/blob/e23ddd5be89ae8712912d4b323db8619ed100de3/examples/asynchronous_publisher_example.py#L319); this seems to assume that the publisher's counter for a given message will always match the delivery tag of that message's ack. I.e. if a message is published with counter 10, any ack with delivery tag ≥ 10 must mean this message is confirmed.

**Reply by lukebakken** [reply]
>  if a message is published with counter 10, any ack with delivery tag ≥ 10 must mean this message is confirmed

Yes that's how it works. You must ack with a tag that has actually been delivered, of course.

**Reply by margaretli27** [reply]
Right. If I understand correctly though, the publisher cannot set the delivery tag (the broker does that). So in the linked example, the publisher increments the message counter but it is not guaranteed that the delivery tag of the message being published exactly matches the current message counter, if e.g. for some reason the publisher restarts and has message counter back at 0. Did I miss anything with that reasoning?

**Reply by lukebakken** [reply]
>  the publisher cannot set the delivery tag (the broker does that)

Are you certain that RabbitMQ sets delivery tags? Check out the `rabbit_channel.erl` source code.
